### PR TITLE
client-side layout re-rendering

### LINKED ISSUE #958: Update Page and Layout Data
Now that all layouts re-render client-side, we need to update the saving/rendering loop to re-enable support for changing component lists in the layout (including changing page data).

## Triggers

* component list dragdrop
* head components drawer dragdrop
* invisible components drawer dragdrop
* component list add / remove
* head components drawer add / remove
* invisible components drawer add / remove

## Saving Data

* saving layout data
* saving page data

## Rendering

* rendering head lists
* rendering body lists (including invisible lists)

### PR DESCRIPTION
**Pull request by nelsonpecora**
fixes #958

* intelligently re-render head components
* intelligently re-render head component lists in the layout
* intelligently re-render body component lists in the layout
* re-render head component lists in the page (from the server)
* re-render body component lists in the page (from the server)

### CODE REVIEW

File: lib/component-data/actions.js [hunk 1]
```diff
@@ -436,7 +444,11 @@ export function openAddComponent(store, { currentURI, parentURI, path, pos }) {
       parentURI,
       path: path,
       components: [{ name: _.head(available) }]
-    }).then((newEl) => store.dispatch('select', newEl)));
+    }).then((newEl) => {
```
Comment (amycheng) [inline lib/component-data/actions.js]: when would `newEl` not be returned?

File: lib/component-data/reactive-render.js [hunk 2]
```diff
@@ -128,25 +127,30 @@ function updateRootDOM(uri, newEl, paths) {
 export function updateDOM(uri, newEl, paths) {
   const isElement = newEl.nodeType === newEl.ELEMENT_NODE,
     isFragment = newEl.nodeType === newEl.DOCUMENT_FRAGMENT_NODE,
-    isDocument = newEl.nodeType === newEl.DOCUMENT_NODE;
+    isDocument = newEl.nodeType === newEl.DOCUMENT_NODE,
+    firstChild = isFragment && newEl.firstElementChild,
+    isFullPageFragment = firstChild && firstChild.tagName === 'META' && firstChild.getAttribute('charset') === 'utf-8';
```
Comment (amycheng) [inline lib/component-data/reactive-render.js]: Isn't this making the assumption that a `layout`-type component will always start out like this? :
```
<!DOCTYPE html>
<html lang="en" data-uri="{{ default _ref _self }}" data-layout-uri="{{ default _ref _layoutRef }}">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
...
```
but I can imagine someone dumping a `<style>` or `<script>` tag before that `<meta>` tag for a special kind of layout.

File: lib/component-data/actions.js (hunk 1 above)
Comment (nelsonpecora) [reply lib/component-data/actions.js]: updating head components, or updating the layout

File: lib/component-data/reactive-render.js (hunk 2 above)
Comment (nelsonpecora) [reply lib/component-data/reactive-render.js]: Hmm, point. Let me see if there's another way to do this.

File: lib/component-data/actions.js (hunk 1 above)
Comment (amycheng) [reply lib/component-data/actions.js]: could you add a code comment here? maybe something like "automatically select the new component if its not a layout or head component"

Comment (nelsonpecora) [reply lib/component-data/actions.js]: added

**Pull request merged**